src/libmongoc/src/mongoc/mongoc-cluster.c

```diff
@@ -3636,3 +3636,99 @@ mcd_rpc_message_decompress_if_necessary (mcd_rpc_message *rpc, void **data, size
 
    return mcd_rpc_message_decompress (rpc, data, data_len);
 }
+
+bool
+mongoc_cluster_run_retryable_write (mongoc_cluster_t *cluster,
+                                    mongoc_cmd_t *cmd,
+                                    bool is_retryable_write,
+                                    mongoc_server_stream_t **retry_server_stream,
+                                    bson_t *reply,
+                                    bson_error_t *error)
+{
+   BSON_ASSERT_PARAM (cluster);
+   BSON_ASSERT_PARAM (cmd);
+   BSON_ASSERT_PARAM (retry_server_stream);
+   BSON_ASSERT_PARAM (reply);
+   BSON_ASSERT (error || true);
+
+   bool ret;
+   // `can_retry` is set to false on retry. A retry may only happen once.
+   bool can_retry = is_retryable_write;
+
+   // Increment the transaction number for the first attempt of each retryable write command.
+   if (is_retryable_write) {
+      bson_iter_t txn_number_iter;
+      BSON_ASSERT (bson_iter_init_find (&txn_number_iter, cmd->command, "txnNumber"));
+      bson_iter_overwrite_int64 (&txn_number_iter, ++cmd->session->server_session->txn_number);
+   }
+
+   // Store the original error and reply if needed.
+   struct {
+      bson_t reply;
+      bson_error_t error;
+      bool set;
+   } original_error = {.reply = {0}, .error = {0}, .set = false};
+
+   // Ensure `*retry_server_stream` is always valid or null.
+   *retry_server_stream = NULL;
+
+retry:
+   ret = mongoc_cluster_run_command_monitored (cluster, cmd, reply, error);
+
+   if (is_retryable_write) {
+      _mongoc_write_error_handle_labels (ret, error, reply, cmd->server_stream->sd);
+      _mongoc_write_error_update_if_unsupported_storage_engine (ret, error, reply);
```

Comment:
Confirming: is inclusion of this call correct in all cases that are now defined in terms of `mongoc_cluster_run_retryable_write()`?

Reply:
I expect "yes". `_mongoc_write_error_update_if_unsupported_storage_engine` implements [this spec requirement](https://github.com/mongodb/specifications/blob/b1e0274ab01183676ec3e21ecb2aa10807107bec/source/retryable-writes/retryable-writes.rst#supported-server-versions):

> Retryable writes are only supported by storage engines that support document-level locking. [...] Drivers MUST raise an actionable error message when the server returns code 20 with errmsg starting with "Transaction numbers".

Before 4d02878f70446968c499230c21f2c13ee0e805d9, `_mongoc_write_error_update_if_unsupported_storage_engine` was only called on the initial attempt (not the retry). If the error is rewritten, I expect no retry is attempted (server does not supporting retryable writes). Therefore, I expect `_mongoc_write_error_update_if_unsupported_storage_engine` is still only called on the initial attempt.